File: packages/php/src/Resources/ContactGroups.php
<?php

declare(strict_types=1);

namespace Opensms\Resources;

use Opensms\Models;
use Opensms\Page;
use Opensms\Transport;

/**
 * The `contactGroups` resource: named groups of contacts, and sending to them. Fields: `name`, `contact_ids`.
 * Accessed as `$client->contactGroups`.
 */
final class ContactGroups
{
    /** @internal */
    public function __construct(private readonly Transport $transport)
    {
    }

    /**
     * Query: `limit` (1..200), `cursor`.
     *
     * @param array<string, mixed> $params
     */
    public function list(array $params = []): Page
    {
        $query = Models::map($params, Models::PAGE_QUERY, [], true);

        return Models::page($this->transport->request('GET', '/v1/contact-groups', $query));
    }

    /**
     * @param array<string, mixed> $params
     * @param array{idempotencyKey?: string} $options
     * @return array<string, mixed>
     */
    public function create(array $params, array $options = []): array
    {
        $body = Models::map($params, Models::CONTACT_GROUP, ['name']);

        return Models::object($this->transport->request('POST', '/v1/contact-groups', [], $body, [
            'idempotent' => true,
            'idempotencyKey' => Models::idempotencyKey($options),
        ]));
    }

    /**
     * @return array<string, mixed>
     */
    public function get(string $id): array
    {
        return Models::object($this->transport->request('GET', '/v1/contact-groups/' . Models::pathParam($id)));
    }

    /**
     * Partial update (PATCH): only the given fields change.
     *
     * @param array<string, mixed> $params
     * @return array<string, mixed>
     */
    public function update(string $id, array $params): array
    {
        $body = Models::map($params, Models::CONTACT_GROUP);

        return Models::object($this->transport->request('PATCH', '/v1/contact-groups/' . Models::pathParam($id), [], $body));
    }

    public function delete(string $id): void
    {
        $this->transport->request('DELETE', '/v1/contact-groups/' . Models::pathParam($id));
    }

    /**
     * Send to every contact in the group. Params: `text` or `template_id`
     * (one of them), `variables` (map), `sender_id`, `traffic_type`,
     * `callback_url`. Returns a Batch that is already running.
     *
     * @param array<string, mixed> $params
     * @param array{idempotencyKey?: string} $options
     * @return array<string, mixed>
     */
    public function send(string $id, array $params, array $options = []): array
    {
        $body = Models::map($params, Models::GROUP_SEND);

        return Models::object($this->transport->request('POST', '/v1/contact-groups/' . Models::pathParam($id) . '/send', [], $body, [
            'idempotent' => true,
            'idempotencyKey' => Models::idempotencyKey($options),
        ]));
    }
}

File: packages/java/src/main/java/io/opensms/ContactGroups.java
package io.opensms;

import io.opensms.models.Batch;
import io.opensms.models.ContactGroup;
import io.opensms.models.Page;

/** Manage contact groups and send to them. Accessed as {@code client.contactGroups()}. */
public final class ContactGroups {

    private final ApiTransport t;

    ContactGroups(ApiTransport t) {
        this.t = t;
    }

    /** @return the first page of groups. */
    public Page<ContactGroup> list() {
        return list(new ListParams());
    }

    /**
     * @param params paging.
     * @return one page.
     */
    public Page<ContactGroup> list(ListParams params) {
        return t.get("/v1/contact-groups", params.toQuery(), ApiTransport.pageOf(ContactGroup.class));
    }

    /**
     * @param params {@code name} required.
     * @return the created group.
     */
    public ContactGroup create(ContactGroupParams params) {
        return create(params, null);
    }

    /**
     * @param params  {@code name} required.
     * @param options per-call options.
     * @return the created group.
     */
    public ContactGroup create(ContactGroupParams params, RequestOptions options) {
        Wire.require(params.name(), "name");
        return t.json("POST", "/v1/contact-groups", params.toWire(), ApiTransport.idempotencyKey(options),
                ApiTransport.type(ContactGroup.class));
    }

    /**
     * @param id group id.
     * @return the group.
     */
    public ContactGroup get(String id) {
        return t.get("/v1/contact-groups/" + Wire.id(id, "id"), null, ApiTransport.type(ContactGroup.class));
    }

    /**
     * @param id     group id.
     * @param params fields to change.
     * @return the updated group.
     */
    public ContactGroup update(String id, ContactGroupParams params) {
        return t.json("PATCH", "/v1/contact-groups/" + Wire.id(id, "id"), params.toWire(), null,
                ApiTransport.type(ContactGroup.class));
    }

    /** @param id group id. */
    public void delete(String id) {
        t.json("DELETE", "/v1/contact-groups/" + Wire.id(id, "id"), null, null, ApiTransport.VOID);
    }

    /**
     * Send to every contact in the group; returns a batch already {@code running}.
     *
     * @param id     group id.
     * @param params text or template.
     * @return the batch.
     */
    public Batch send(String id, GroupSendParams params) {
        return send(id, params, null);
    }

    /**
     * @param id      group id.
     * @param params  text or template.
     * @param options per-call options.
     * @return the batch.
     */
    public Batch send(String id, GroupSendParams params, RequestOptions options) {
        return t.json("POST", "/v1/contact-groups/" + Wire.id(id, "id") + "/send", params.toWire(),
                ApiTransport.idempotencyKey(options), ApiTransport.type(Batch.class));
    }
}
